### ISSUE TITLE
1.6: The files/config_.../active" configuration must be completely empty to proceed with installation.

### ISSUE THREAD

**Issue by DrupaListo-com** [backdrop/backdrop-issues#2507]
I just tested installing backdrop 1.6 for the first time - quite disappointed because it gave me a stupid error while installing: "The files/config_.../active" configuration must be completely empty to proceed with installation.

Can somebody test a new install and confirm?

**Comment by Dinornis** [comment]
I think this may be being addresses in https://github.com/backdrop/backdrop-issues/issues/2263
Marking as duplicate.

**Comment by olafgrabienski** [comment]
@Dinornis - Not sure if it's a duplicate. @DrupaListo-com said he installed Backdrop for the first time while #2263 seems to happen when you *reinstall* a site.

@DrupaListo-com Can you tell if there is something in the mentioned directory (and what)?

**Comment by serundeputy** [comment]
Hello @DrupaListo-com thanks for reporting!

I just tested installing a fresh `1.6.0` no errors.  Installation went fine.
As @olafgrabienski can you give us some additional info about the steps you are taking so that we can replicate and solve this problem?

In my test I installed in english (in case that matters).

thanks,
~Geoff

**Comment by klonos** [comment]
Same as the others here @DrupaListo-com I have installed quite a few test environments of 1.6.0 since its release a few weeks ago and I never had an issue.

**Comment by jenlampton** [comment]
Maybe @DrupaListo-com had installed Backdrop previously, but forgot. I often forget to empty my config directory when reinstalling.

**Comment by DrupaListo-com** [comment]
Hi guys, I am a long time (10 years) drupal-er and I have installed 100-s of sites.
Pretty sure it was an empty folder.
Stay tuned for my second fresh test...

**Comment by jenlampton** [comment]
Tha for trying again @DrupaListo-com, I often forget to empty out the config directory because that's not something that was required with Drupal. 

Also if it installed half-way and failed (like on the requirements page) it could have been that some files were written into that config directory, and it threw itself off.  I've noticed that happening sometimes, but haven't been able to document exactly what circumstances cause it.

**Comment by DrupaListo-com** [comment]
My 2nd attempt was unsuccsessfull, too.
Perhaps I am doing something a little bit exotic and that's why. But such behaviour is definitely a bug.

@jenlampton I think I have several possible explanations as to why things get broken with my install and its similar to your explanation above:
- some errors on mysql settings page:
  - wrong mysql password or wrong mysql host: localhost vs 127.0.0.1 - depending on mysql install one/both will work and in case only 127.0.0.1 would work but the default is "localhost", mysql can't connect
  - so I think if sth is wrong and u submit the form, the install populates the config dir but stops in the middle and later checks if dir is empty and fails ...
- I am using a db prefix - I think that as well might be a problem somehow.

I will investigate further and report more findings. 

Anyone is free to test a bit the install process and try to break it with by a) giving wrong mysql pass/user/host data and b) using some db table prefix.

Whoever can replicate first will be my hero.

**Comment by Graham-72** [comment]
@DrupaListo-com I am having no problem installing Backdrop 1.6

However, in the past I have had problems installing Backdrop on servers even when Drupal will install. This is something to do with the details of the PHP configuration on the server. Are you able to try a different version/configuration of PHP? See comments on issues #2204 and #590 for more detail.

**Comment by serundeputy** [comment]
yep; thanks @DrupaListo-com the missing piece was datbase prefix.

* download backdrop
* visit {your_site}/install.php
* On the database credentials page open the advanced fieldset
* fill in a database prefix
* you will receive an error:

![screen shot 2017-02-05 at 2 49 23 pm](https://cloud.githubusercontent.com/assets/6763671/22629389/e2aa42ea-ebb2-11e6-9620-c2f6ce27f5cf.png)

At this point the installer has created the `flles/config_{hash}` directory:

<img width="566" alt="screen shot 2017-02-05 at 2 50 09 pm" src="https://cloud.githubusercontent.com/assets/6763671/22629393/ff077732-ebb2-11e6-8c24-43a267194951.png">

If you remove this directory:

```
rm -rf files/config_{hash}
```

and then refresh the installer from the error page the installation will complete, but you have some leftover initial tables without the db_prefix and the corresponding prefixed table.  For example there will be a `cache` table and  `db_prefix_cache` table.

<img width="1680" alt="screen shot 2017-02-05 at 2 51 05 pm" src="https://cloud.githubusercontent.com/assets/6763671/22629404/4d131c38-ebb3-11e6-9363-13ce1a3fa286.png">

So we need to fix the installer to handle `db_prefix` option correctly.

Thanks for staying with it @DrupaListo-com

**Comment by serundeputy** [comment]
For what it is worth I downloaded `1.5.2` and `1.0.0` and they both suffer from the same bug ... so it seems likely that `db_prefix` was never working and not something introduced in `1.6.0` that broke it.

**Comment by DrupaListo-com** [comment]
@serundeputy exactly what I experienced! Was too busy/lazy to report with shots and details as u did. Kudos!

**Comment by jenlampton** [comment]
I've updated the issue title and added the 1.9.6 milestone, as this is a fairly major bug. I'd love to see it resolved in 1.9.6. Anyone have any idea why database prefixes no longer work? Maybe it's because we went back to the Drupal 6 style db connection string, and didn't update database prefix handling accordingly?

**Comment by quicksketch** [comment]
We use DB prefixes when running tests, and I think they're used when spinning up the sandboxes on PRs as well, so generally we know for certain that DB prefixes are actually still supported by the entire application.

This problem seems to be specific to just the UI installer form. If you specify a `$database_prefix` in settings.php, it installs just fine. The problem simply seems to be that the submit handler assumes that the database settings are keyed the same as a `$database` array, but the prefix field is `db_prefix` instead of `prefix`. Changing this key fixes the form, and the database settings get processed correctly.

PR at https://github.com/backdrop/backdrop/pull/2181.

**Comment by herbdool** [comment]
I checked Drupal 7 to see if it had the same error. I saw this in includes/install.core.inc:

```
 // TODO: remove when PIFR will be updated to use 'db_prefix' instead of // 'prefix' in the database settings form. $database['prefix'] = $database['db_prefix']; unset($database['db_prefix']); 
```

I don't know what PIFR is but looks like they just made the two keys equal in the validation.

**Comment by quicksketch** [comment]
Yeah, PIFR is part of drupal.org's testing automation (aka testbot). If that's the only limiting factor, then I think we're set here. Our testing suite uses the `/core/scripts/install.sh` script to automate installs, which is updated as part of this PR as well.
